### ISSUE TITLE
Use Math.imul(a, b) instead of a * b | 0

### ISSUE THREAD

**Issue by rightfold** [purescript/purescript#2980]
`Math.imul` correctly handles arithmetic overflow, whereas `| 0` just discards the most significant bits. This affects purescript-prelude as well as the inliner.

**Comment by garyb** [comment]
It'd be better if we can auto-polyfill this or something, since `imul` does not have solid browser support. It's not even ES5, nevermind the currently-supported ES3. 😕

**Comment by hdgarrood** [comment]
There's already an issue for this, isn't there?

**Comment by hdgarrood** [comment]
Ok, there's https://github.com/purescript/purescript/issues/2921, but it was closed, so maybe let's just use this one.

MDN's [Math.imul page](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/imul) suggests the following polyfill:
```javascript
Math.imul = Math.imul || function(a, b) {
  var ah = (a >>> 16) & 0xffff;
  var al = a & 0xffff;
  var bh = (b >>> 16) & 0xffff;
  var bl = b & 0xffff;
  // the shift by 0 fixes the sign on the high part
  // the final |0 converts the unsigned value into a signed value
  return ((al * bl) + (((ah * bl + al * bh) << 16) >>> 0)|0);
};
```

**Comment by hdgarrood** [comment]
Come to think of it, we shouldn't mutate `Math`, and we can't exactly inline all that every time someone uses Int's `*`. Without a stdlib it's hard to see how to achieve this without either giving up inlining or dropping support for platforms which don't provide `Math.imul`.

**Comment by hdgarrood** [comment]
If we had a general inliner i.e. one that didn't only use a set of special cases built in to the compiler, this would probably be doable, though.

**Comment by michaelficarra** [comment]
@hdgarrood Do we have an official stance on JS engine support? It'd be nice to make a requirement that can shift over time like "browsers with >0.1% market share (by page loads) globally" or "browsers with >0.5% market share (by installations) in every individual market" or "browsers released in the last 5 years". A target of "just ES3" is problematic both because it doesn't actually represent meaningful browser support, and it doesn't allow us to improve/adapt as time passes and things get better.

**Comment by natefaubion** [comment]
I think it's reasonable to say that PureScript does not require a runtime, but relies on some set of easily polyfilled features, such as `Math.imul` or `Object.keys`.

**Comment by hdgarrood** [comment]
I don't think we really have an official stance. The compiler generates ES3, most of the Node libraries assume node >= 4, and I guess from @natefaubion's comment that maps assumes Object.keys, so there's a mix. I think what you say makes sense though. I guess we haven't really needed to develop a clear stance as I don't think there are that many situations where we can't reasonably accomplish what we want using just ES3.

Would we be able to justify using Math.imul with any of those stances you gave, incidentally?

**Comment by Pauan** [comment]
@hdgarrood I personally don't see a problem with mutating `Math`

The polyfill will only be added if `Math.imul` doesn't already exist, and the polyfill should have the correct behavior (it's just slower than native support), so I don't think the polyfill will cause any problems.

@michaelficarra [`Math.imul` is supported in every browser except IE11.](https://kangax.github.io/compat-table/es6/#test-Math_methods_Math.imul_a_href=_https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/imul_title=_MDN_documentation_img_src=_../mdn.png_alt=_MDN_(Mozilla_Development_Network)_logo_width=_15_height=_13_/_/a_nbsp;)

This is because `Math.imul` is super easy for browsers to support (it's just standard C multiplication), and it was necessary for asm.js support, so all the browsers added it in really fast.

IE11 has ~4% market share, so I don't think it's viable to drop support for it right now.

**Comment by paf31** [comment]
Is there anything wrong with providing this as a library function in `math` for now?

**Comment by garyb** [comment]
Only that it doesn't fix `(*)` for integers.

**Comment by hdgarrood** [comment]
If we mutated `Math` I reckon that in 99% of projects it would work fine, but I'm certain there will be people who have some weird setup that makes strange decisions based on the value of this property and would send them (and possibly also us) on a debugging goose chase for hours. It's maybe justifiable to mutate globals if you're writing a framework but I think it's poor etiquette to do so in a library. I just remember too many instances of this kind of code breaking things in totally unpredictable and unexpected ways back when I was an Rubyist.

If we were going to provide a temporary opt-in solution I'd rather do it in the form of a newtype over Int than a function in `math`, as I think that would be more useful. But what would we be waiting for if we did that? For IE's market share to die off sufficiently so that we could assume the presence of Math.imul?

**Comment by paf31** [comment]
It wouldn't fix `(*)` but I think that might be okay. We could just document that `Int` works most of the time but that there are catches, and that you can choose to use `imul` via a `newtype` if you want the fix at the cost of compatibility.

**Comment by hdgarrood** [comment]
That sounds like a good thing to do for now — the `Semiring` documentation is already sufficient in that sense I think. I am still quite keen to get the default `Semiring Int` instance fixed, though.

**Comment by Pauan** [comment]
@hdgarrood If some code breaks because `Math.imul` exists, then it's *already* broken on all modern browsers. The only possible way it could break is if it relied upon the string representation of the `Math.imul` function (which is a *horrible* thing to do in any case).

Yes global mutation has been abused in the past in JavaScript, and of course I recommend against it, but in *this specific* case, it should be fine.

Having said that, I'm okay with putting it in `math` and using a `newtype`

**Comment by rightfold** [comment]
@pauan you can overwrite `toString` :)

**Comment by hdgarrood** [comment]
I'm not convinced, personally - I've been bitten too many times by conflating "this is safe" with "I cannot think of a way this could be unsafe". And it's not just about that property existing - it seems a reasonable assumption to make that the number of properties of `Math` isn't going to change over the course of running the code on a webpage, for instance, but this would no longer be true. It probably only is horrible code which could be affected but the quality of any other code on a page containing some PureScript code isn't any of our business, imo.

**Comment by rightfold** [comment]
If we document that overflow on multiplication gives an unspecified result then switching to `Math.imul` later won't be a breaking change.

**Comment by hdgarrood** [comment]
Ok, so to recap: since no version of IE supports `Math.imul`, it will probably be a quite long time before we can just start using it without having to worry about compatibility.

We could say that PureScript depends on some easily-polyfillable features, and just let the user polyfill themselves if they want IE support, but for something as basic as integer multiplication I think this would be a poor UX for devs not familiar with PureScript and/or this specific issue.

We could polyfill it ourselves by getting the compiler to insert some code which mutates `window.Math` to add the `imul` property if there are any uses of multiplication with the `Int` type, but I'd rather not do this because I think it is poor etiquette. It also sounds like it could be quite fiddly, and I think it goes against the "no stdlib" principle.

One alternative option that has just occurred to me is that we change the definition of `intMul` so that it uses `Math.imul` via the FFI if it is there, or otherwise it uses a polyfill, but whichever path is taken, it should be done without mutating `window.Math`. See [this PR](https://github.com/purescript/purescript-math/pull/18) for how we might achieve that (thanks @matthewleon :smile:). Then, we'd need to improve the inliner so that this could be inlined properly (i.e. by generating a `require` statement if necessary). I think this would be my preferred option, although I'm not sure how difficult modifying the inliner would be.

**Comment by paf31** [comment]
Another option is to make it explicit that the user needs to polyfill things, either by moving it into its own module (`Math.NonPortable` or something), or by using a type class à la `Partial`.

**Comment by LiamGoodacre** [comment]
Anything bad about changing the definition of `intMul` https://github.com/purescript/purescript-prelude/blob/v4.1.0/src/Data/Semiring.js#L10 to be the defaulted polyfill that @hdgarrood posted?
```javascript
exports.imul = Math.imul || function(a, b) {
  var ah = (a >>> 16) & 0xffff;
  var al = a & 0xffff;
  var bh = (b >>> 16) & 0xffff;
  var bl = b & 0xffff;
  // the shift by 0 fixes the sign on the high part
  // the final |0 converts the unsigned value into a signed value
  return ((al * bl) + (((ah * bl + al * bh) << 16) >>> 0)|0);
};
```
```purescript
foreign import imul :: Fn2 Int Int Int
instance semiringInt :: Semiring Int where
  add = intAdd
  zero = 0
  mul = runFn2 imul
  one = 1
```

**Comment by LiamGoodacre** [comment]
Ah sorry @hdgarrood didn't read you latest comment properly, you already suggested that ^

**Comment by paf31** [comment]
Inlining?

**Comment by LiamGoodacre** [comment]
I guess we would inline from `x * y` to `runFn2 imul x y`?

**Comment by paf31** [comment]
That sounds like quite the deoptimization for quite a lot of simple numeric code.

**Comment by LiamGoodacre** [comment]
I've set up a jsperf (https://jsperf.com/imul-or-mul).  Unless it's a bad test, I get unexpectedly comparable performance between the following on Chrome:
```
(x * y) | 0
Math.imul(x | 0, y | 0) | 0
imulPolyfill(x | 0, y | 0) | 0
```

**Comment by JordanMartinez** [comment]
![firefox62-linuxmint18-3](https://user-images.githubusercontent.com/8413037/46103805-9ab2fd00-c186-11e8-9bad-faaf8b7b9339.png)

**Comment by LiamGoodacre** [comment]
Perhaps a slightly better structure of test? https://jsperf.com/mul-imul-poly-2

But yeah, under the assumption that any of these tests are useful, it seems to me that `imulPolyfill(x | 0, y | 0) | 0` would be an acceptable choice of implementation.

**Comment by iamahuman** [comment]
There's an alternative implementation from MDN (and of course we are not mutating `Math`; this is just verbatim):

```js
if (!Math.imul) Math.imul = function(opA, opB) {
  opB |= 0; // ensure that opB is an integer. opA will automatically be coerced.
  // floating points give us 53 bits of precision to work with plus 1 sign bit
  // automatically handled for our convienence:
  // 1. 0x003fffff /*opA & 0x000fffff*/ * 0x7fffffff /*opB*/ = 0x1fffff7fc00001
  //    0x1fffff7fc00001 < Number.MAX_SAFE_INTEGER /*0x1fffffffffffff*/
  var result = (opA & 0x003fffff) * opB;
  // 2. We can remove an integer coersion from the statement above because:
  //    0x1fffff7fc00001 + 0xffc00000 = 0x1fffffff800001
  //    0x1fffffff800001 < Number.MAX_SAFE_INTEGER /*0x1fffffffffffff*/
  if (opA & 0xffc00000 /*!== 0*/) result += (opA & 0xffc00000) * opB |0;
  return result |0;
};
```

Those polyfills would basically only target IE, and its JScript engine wouldn't perform any better than this on the more widely known one unless the engine has any optimizations for integer arithmetic, which seems unlikely.

@LiamGoodacre I think the `Math.imul` option needs to be reevaluated with a correction to fetch the `imul` property and store it in a closure beforehand, pulling the property get overhead out of the loop. And it's what we're going to do anyway, since that way `(*)` would preserve its referential transparency even if later it's set to any other value. That is, assuming the property is kept its initial value upon PureScript's runtime initialization.

**Comment by natefaubion** [comment]
Is there anything to resolve here? I think we should make a decision one way or the other.

**Comment by hdgarrood** [comment]
I think we should definitely fix Int multiplication one way or another. My preference is this:

 > One alternative option that has just occurred to me is that we change the definition of intMul so that it uses Math.imul via the FFI if it is there, or otherwise it uses a polyfill, but whichever path is taken, it should be done without mutating window.Math. See [this PR](https://github.com/purescript/purescript-math/pull/18) for how we might achieve that (thanks matthewleon). Then, we'd need to improve the inliner so that this could be inlined properly (i.e. by generating a require statement if necessary). I think this would be my preferred option, although I'm not sure how difficult modifying the inliner would be.

**Comment by natefaubion** [comment]
I don't really understand the "generating a require statement if necessary". It isn't clear to me (based on the jsperfs) that we need to do anything other than just disable the inliner for Int mul. I would expect that in hot code it would get inlined anyway by the JIT.

**Comment by hdgarrood** [comment]
I was hoping that we would be able to retain inlining for Int mul; I don't think the jsperfs necessarily demonstrate that we can justify disabling the inliner? That is,  `x * y | 0` may be similar to `Math.imul(x,y)` performance-wise but we really ought to be comparing it to `Data_Semiring.semiringInt.mul(x)(y)` where `Data_Semiring.semiringInt.mul = function(x) { return function(y) { return Math.imul(x,y) } }` right (or perhaps something with even more indirection if the Semiring instance is coming from a chain of subclasses)?

At least there doesn't appear to be very much of a difference between the curried version of Math.imul versus `*`: https://jsperf.com/mul-imul-curried/1. Do you have a feel for how costly the object property lookups are likely to be hot code, particularly in the case where type class hierarchies are involved?

**Comment by natefaubion** [comment]
Are type class hierarchies relevant? The inline rule will only trigger on monomorphic instantiations, so I don't think a hierarchy would get involved. By generating a `require`, do you mean importing `Data.Int` and replacing the call with something like `runFn2 Data_Int.mulFn2 a b`?

**Comment by natefaubion** [comment]
I would expect the object property lookups to be inline-cached since they are not from a dynamically created instance.

**Comment by hdgarrood** [comment]
> Are type class hierarchies relevant? The inline rule will only trigger on monomorphic instantiations, so I don't think a hierarchy would get involved.

Oh of course yes, I'd forgotten about this.

> By generating a require, do you mean importing Data.Int and replacing the call with something like runFn2 Data_Int.mulFn2 a b?

Yes. But given what you've pointed out here, I am happy to try simply disabling the inliner to start with. We can always revisit if it does turn out to be a problem. Maybe we can release the change to the instance in `prelude` as a breaking change alongside 0.14.0.

**Comment by natefaubion** [comment]
I mean, I do think we should benchmark real output from the compiler before merging anything.

**Comment by rhendric** [comment]
So, does anyone have a PS project that makes sufficiently heavy use of integer multiplication that it would be a good benchmark for this change?

**Comment by JordanMartinez** [comment]
I'm stating my understanding of the current issue since it's spread across multiple comments now.

It sounds like we want to use `Math.imul` rather than `a * b | 0` to do integer multiplication because the former is believed to be more performant than the latter. While using `Math.imul` is desirable, the issue is that IE11 does not have the `imul` function defined on the global `Math` object. Moreover, mutating the `Math` object to include that just goes against principle. 

Thus, the current proposal is to use `Math.imul` if it exists or default back to a polyfill. However, this proposed solution may incur overhead, so a benchmark test is needed to verify what affect such a change has upon performance. Moreover, it sounds like we do not currently have a good project with which to benchmark the changes.

With that being said, if IE11 wasn't a factor, would we default to using `Math.imul`? Or is the benchmark still necessary even after that?

And just a few other updates about and questions regarding IE11.

First, [IE11 will no longer be supported on _some_ versions of Windows, starting June 15, 2022](https://docs.microsoft.com/en-us/lifecycle/faq/internet-explorer-microsoft-edge#is-internet-explorer-11-the-last-version-of-internet-explorer-) (i.e. ~3 months away), but Edge will have an "Internet Explorer Mode" that enables this legacy support until at least 2029.

Second, I'm not sure if Edge-with-IE-mode-enabled traffic is distinguished from just-Edge traffic. I assume not, so these stats on the browser market share regarding IE11 might not be as helpful:
- [`1.45%` in 2021 via backlinko.com](https://backlinko.com/browser-market-share#desktop-browser-share)
- [`1.31%` as of Dec 2021 via x](https://www.statista.com/statistics/272697/market-share-desktop-internet-browser-usa/)
- [`0.6%` as of Feb 2022 via statcounter.com](https://gs.statcounter.com/browser-market-share)

Third in #4232 and #4250, all FFI must be ES modules; a CommonJS module produces a failure. A future PR will also remove `purs bundle`. So, will it be possible to target IE11 in PureScript `v0.15.0` without using Babel or any other such tooling to add any needed polyfills automatically?

Fourth, is our concern here more about the fear of preventing the existing PS community users from targeting IE11? Or just the general principle of supporting as many browsers as possible? In other words, when the survey results come out, if no one who is currently writing PS on a regular-ish basis is targeting IE11, is that a strong enough reason to no longer support IE11 officially, even if some people who do not write PS regularly but still filled out the survey target IE11 through some other language?

**Comment by garyb** [comment]
> It sounds like we want to use `Math.imul` rather than `a * b | 0` to do integer multiplication because the former is believed to be more performant than the latter.

Not quite, the main motivation is because `Math.imul` handles integer overflow correctly where `a * b | 0` does not. It would make `Int` a fully law-abiding `Semiring`.

**Comment by hdgarrood** [comment]
If we were inlining `Int` multiplication to `Math.imul`, we wouldn't need to benchmark. The purpose of the benchmark is to measure the performance impact of no longer inlining `a * b`, and instead having it compile to a curried function, which is what this proposal likely involves.

I think the ES modules discussion is irrelevant - ES modules shouldn't result in us dropping support for any browsers that we currently support, because the idea is that you'll bundle before shipping to browsers, and the generated JS should look pretty much the same as it did before ESM if you're bundling.

In general, the further up some component is in the ecosystem, the more careful we need to be about dropping support for browsers/environments/whatever in that component, because the more people there are likely using the component, and the harder it'll be for people to switch out the component for an alternative component which does support all of the browsers they're targeting. The compiler is right at the top - everything in the ecosystem depends on the compiler - so we need to apply this principle most strictly here. 1% of people is quite a lot of people; breaking an app that people rely on, like, say, their email is unacceptable. It's often possible to upgrade but not always - for example, the effort to move off of Windows XP in the UK's National Health Servic was still ongoing of 2019 and might not be finished even now - and I don't think it's fair to punish users just because the people who manage their IT equipment are inept. I think for it to be acceptable for us to drop support for IE11, there has to be no good way of continuing to support IE11. As far as I can tell, that is not the case here.

**Comment by natefaubion** [comment]
> I think for it to be acceptable for us to drop support for IE11, there has to be no good way of continuing to support IE11. As far as I can tell, that is not the case here.

As stated above, there are polyfills for this. Continuing to support IE11 would entail users who want that support to add a `Math.imul` polyfill which is super easy. It can even be emitted as part of a bundle or your PS app entry point. Why is that not acceptable?

**Comment by natefaubion** [comment]
FWIW, I'm personally in favor of requiring and documenting polyfills for select features for old browser support. I do not agree that we must have non-polyfilled support for IE11.

**Comment by michaelficarra** [comment]
@natefaubion Why would performance differ?

```js
export const imul = Math.imul || function(a, b) { /* ... */ };
```

The switch on its presence happens only once.

**Comment by natefaubion** [comment]
> @natefaubion Why would performance differ?

Per @hdgarrood 

> I was hoping that we would be able to retain inlining for Int mul; I don't think the jsperfs necessarily demonstrate that we can justify disabling the inliner?

I already expressed skepticism that there would be a performance hit, however nobody has presented a benchmark to test this. So without out we can only rely on the reasoning so far, and I otherwise don't know how to move past @hdgarrood's objections. Polyfills are super common, and most definitely so if you are targeting IE11. To me that seems like the most straightforward way to move past it without regressions.

**Comment by JordanMartinez** [comment]
> FWIW, I'm personally in favor of requiring and documenting polyfills for select features for old browser support. I do not agree that we must have non-polyfilled support for IE11, especially if it means better correctness and performance for 98.5% of users.

This is where I stand, too.

**Comment by hdgarrood** [comment]
The reason I was imagining there could be a performance hit was that:
- Inlining of Int multiplication after this change would mean that `a * b` should become `runFn2 intMul a b`, where `intMul` would be `Math.imul` in environments that have it, and a polyfill otherwise
- assuming we aren't mutating `Math`, `intMul` would have to be defined in some other PureScript module, which would mean that the act of inlining might require us to add another import to the current module
- Our inliner is not currently capable of this and it's probably a fair bit of effort to add
- So we'd probably not bother, which would mean that `a * b` would compile to something like `Data_Semiring.semiringInt.mul(a)(b)`, which I expect would be slower than `Math.imul(a, b)` because of the currying.

Of course this concern is moot if we do mutate `Math`, or if we assume the presence of `Math.imul` and document that users need to mutate `Math` to add `imul` if it's not there, because in that case we can just inline Int multiplication to `Math.imul` directly (and I'd not feel any need to benchmark if we were doing that).

I'll withdraw my objection to polyfilling, I agree that it's easy enough to do and it's true that people who are targeting IE11 are likely doing some of it already anyway. I don't have a strong preference on whether we do it automatically or whether we ask the user to do it themselves.

The thing that bothers me is the suggestion that if if few people who are currently writing PS are consciously targeting IE11, then that alone represents a reason to drop support for it - I think it's really important to bear in mind a) the possibility that people who are writing PureScript have IE11 users anyway, regardless of whether they are consciously supporting those users, and b) that low usage alone is never enough to consider dropping support for something; imo an argument to drop support for something should _always_ include not only figures demonstrating that it's rarely used, but also examples of how dropping it improves things for everyone else in a way that we couldn't reasonably achieve otherwise.

**Comment by natefaubion** [comment]
> imo an argument to drop support for something should always include not only figures demonstrating that it's rarely used, but also examples of how dropping it improves things for everyone else in a way that we couldn't reasonably achieve otherwise.

This is pretty much an impossibly high barrier given our resources, and is essentially saying "we always have to support IE11" because no one is reasonably going to provide those numbers. You can say the same for the inverse: If it concerns you that much, then you should prove that there are a lot of IE11 users it affects. I personally don't think either is a given. Note: you also have not stated why 1% is "a lot of users" or what would constitute "not a lot of users", because realistically neither side has any idea if those users and PureScript consumers overlap. So it feels like this just pushes the goal posts indefinitely.

**Comment by garyb** [comment]
I could be wrong, but I think the intention was the "local polyfill" version of `imul` wouldn't be inserted as `runFn2 intMul a b`, it'd be inserted as `imul(x, y)`, just like it was a call to `Math.imul` / the current operation. I guess we'd call it `$imul` or something to avoid name collisions too.

There is a some additional complexity there, in that we'd need to insert it in any module where that inline was made (or have some other means of referencing/importing a global version of it), but anyway, my impression was it'd be handled at the JS level rather than the PS level, so no `runFn2`.

**Comment by natefaubion** [comment]
I think that this thread is relevant then:
https://github.com/purescript/purescript/issues/4179#issuecomment-1056021493

This to me would constitute another example where a small JS runtime module would be the most straightforward "best of all worlds" solution.

**Comment by rhendric** [comment]
> wouldn't be inserted as `runFn2 intMul a b`, it'd be inserted as `imul(x, y)`

Is there a difference? Don't we already inline `runFn2(intMul)(a)(b)` to `intMul(a, b)` in the CoreImp optimizer?

**Comment by hdgarrood** [comment]
> This to me would constitute another example where a small JS runtime module would be the most straightforward "best of all worlds" solution.

Agree tbh, I also think the "no runtime" claim is fairly dubious.

**Comment by garyb** [comment]
> Is there a difference? Don't we already inline `runFn2(intMul)(a)(b)` to `intMul(a, b)` in the CoreImp optimizer?

Good point

> > This to me would constitute another example where a small JS runtime module would be the most straightforward "best of all worlds" solution.
> 
> Agree tbh, I also think the "no runtime" claim is fairly dubious.

Yeah, same on both counts. I'm not sure the things we're proposing are even what I'd call a runtime either.

**Comment by natefaubion** [comment]
> I'm not sure the things we're proposing are even what I'd call a runtime either.

Right, this is why I consider the claim dubious. I think definitions here vary. I personally consider "no runtime" to mean "no special calling convention", which neither proposal violates. However, taken at face value, it can also mean "no importing files that the user didn't write", which I don't agree with.

**Comment by f-f** [comment]
One year since the last comment, the market share for IE is now [0.24% according to statcounter](https://gs.statcounter.com/browser-market-share).
Can we just switch to `imul` and resolve this?

**Comment by Qqwy** [comment]
If we do want to support the polyfill on ancient browsers, wouldn't the following tiny change to https://github.com/purescript/purescript-prelude/blob/f4cad0ae8106185c9ab407f43cf9abf05c256af4/src/Data/Semiring.js#L8-L13C3 be enough?

```js
export const intMul = Math.imul ? function (x) { return function (y) { return Math.imul(x, y); }; } : function (x) {
  return function (y) {
    /// The polyfill from https://github.com/purescript/purescript/issues/2980#issuecomment-495628586
    opB |= 0;
    var result = (opA & 0x003fffff) * opB;
    if (opA & 0xffc00000 /*!== 0*/) result += (opA & 0xffc00000) * opB |0;
    return result |0;
  };
};
```

**Comment by garyb** [comment]
I'm in favour of switching now. Given we now have ES modules and support stuff like arrow syntax, I would imagine most projects already won't work in IE without extra tooling and polyfills.

**Comment by MonoidMusician** [comment]
@Qqwy The main issue isn't the binding in the library, the issue is that it also needs to be supported by the inliner in some manner (since we want primitives on `Int` to be efficient).
